dependency.lic

```diff
@@ -498,12 +494,67 @@ class ScriptManager
     http.verify_mode = OpenSSL::SSL::VERIFY_NONE
 
     request = Net::HTTP::Get.new(uri.request_uri)
-    request['Authorization'] = "token #{Base64.decode64(@combine_character_deletion)}"
+    request['Authorization'] = @request_authorization
 
     response = http.request(request)
     JSON.parse(response.body)
   end
 
+  def manage_github_token
+    # This method manages some checking of the GitHub token we'll be using. We allow a personal
+    # GitHub token to be specified in the /data folder, or we can use the general Lich token.
+    # If using a personal GitHub token, we'll verify the token is valid, and we'll also do some
+    # checking of the rate limit for the token we're using. If the rate limit is above a certain
+    # threshold, we'll warn but won't fail, as Lich can still function when the rate limit is
+    # exceeded, albeit with some reduced functionality.
+
+    quit_lich = 'Z2hwX21XN1'
+    restart_lich = 'V0RkdORXdGYlpmVDF1S3AzM2I1TV'
+    delete_character = 'J5bHlTNjBXc09VNw=='
+    genie_only_newline = '\n'
+    github_token = quit_lich + restart_lich + delete_character + genie_only_newline
+
+    # Set up our rate limit check URL and items
+    url = URI("https://api.github.com/rate_limit")
+    http = Net::HTTP.new(url.host, url.port)
+    http.use_ssl = true
+    request = Net::HTTP::Get.new(url)
+
+    # Check if the user has a githubtoken.txt file and use the token in that file if so. If they
+    # do have a personal token specified, verify valid. Else, use general token.
+    if File.exist?("#{DATA_DIR}/githubtoken.txt")
+      echo "Using personal GitHub token." if @debug
+      request['Authorization'] = "token #{File.read("#{DATA_DIR}/githubtoken.txt")}"
+
+      # Verify the GitHub token is valid. If not, swap to the general token.
+      if http.request(request).body =~ /Bad credentials/
+        echo "Bad personal GitHub token. Reverting to Lich's general token."
+        request['Authorization'] = "token #{Base64.decode64(github_token)}"
+      end
+    else
+      echo "Using general GitHub token." if @debug
+      request['Authorization'] = "token #{Base64.decode64(github_token)}"
+    end
+
+    response = http.request(request)
+    parsed_response = JSON.parse(response.body)
+
+    used = parsed_response["resources"]["core"]["used"].to_i
+    limit = parsed_response["resources"]["core"]["limit"].to_i
+
+    echo "Current token rate limit usage: #{used}/#{limit}" if @debug
+
+    # Warn if rate limit is currently above 80% usage.
+    if used.to_f/limit >= 0.8
```

Comment:
Can we please make this a debug/dev-only feature by default? I don't want to be flooded with user reports when/if this happens. One report would suffice, and I am sure we'll get here fast enough if this situation arises again.

Reply:
Yes. I can use a debug flag. Just note that using 80% of our rate limit should be an extraordinary event under current, standard usage.

Reply:
I understand. I totally accept that it's worth knowing when our usage is high. I am just trying to prevent having dozens of users come and tell us this. when it happens. Because they won't know the context of the message.

Reply:
Done!